src/Razor/src/Microsoft.VisualStudio.Editor.Razor/Snippets/XmlSnippetParser.CodeSnippet.cs

```diff
@@ -0,0 +1,112 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Xml.Linq;
+using Microsoft.AspNetCore.Razor;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.Editor.Razor.Snippets;
+
+internal partial class XmlSnippetParser
+{
+    /// <summary>
+    /// Shamelessly copied from the editor
+    /// https://devdiv.visualstudio.com/DevDiv/_git/VS-Platform?path=/src/Editor/VisualStudio/Impl/Snippet/CodeSnippet.cs
+    /// </summary>
+    internal class CodeSnippet
+    {
+        private const string ExpansionSnippetType = "Expansion";
+
+        private readonly string[]? _snippetTypes;
+
+        /// <summary>
+        /// Ctor.
+        /// </summary>
+        /// <param name="codeSnippetElement">XElement representing the CodeSnippet node.</param>
+        public CodeSnippet(XElement codeSnippetElement)
+        {
+            var header = GetElementWithoutNamespace(codeSnippetElement, "Header");
+            if (header == null)
+            {
+                throw new InvalidOperationException("snippet element is missing header.");
+            }
+
+            CodeSnippetElement = codeSnippetElement;
+
+            Title = GetElementInnerText(header, "Title");
+            Shortcut = GetElementInnerText(header, "Shortcut");
+            var snippetTypes = GetElementsWithoutNamespace(header, "SnippetTypes");
+            if (snippetTypes != null)
+            {
+                _snippetTypes = snippetTypes.Elements().Select(e => e.Value.Trim()).ToArray();
+            }
+        }
+
+        public string Title { get; }
+
+        public string Shortcut { get; }
+
+        public XElement CodeSnippetElement { get; }
+
+        public bool IsExpansionSnippet()
+        {
+            return _snippetTypes?.Contains(ExpansionSnippetType, StringComparer.OrdinalIgnoreCase) == true;
+        }
+
+        public static CodeSnippet ReadSnippetFromFile(string filePath, string snippetTitle)
+        {
+            var document = XDocument.Load(filePath);
+            var snippets = ReadSnippets(document);
+            snippets.AssumeNotNull();
```

Comment:
Can be removed if you take the other suggestions to avoid nullable ImmutableArray.